### ISSUE TITLE
Spring boot actuator - prometheus metrics not working

### ISSUE THREAD

**Issue by radut** [prometheus/client_java#215]
Hi,
Spring boot prometheus endpoint is not working.

in target tabs of prometheus i have  error : text format parsing error in line 1: invalid metric name

```
curl -v http://192.168.1.158:8083/actuator/prometheus
*   Trying 192.168.1.158...
* TCP_NODELAY set
* Connected to 192.168.1.158 (192.168.1.158) port 8083 (#0)
> GET /actuator/prometheus HTTP/1.1
> Host: 192.168.1.158:8083
> User-Agent: curl/7.53.1
> Accept: */*
>
< HTTP/1.1 200
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: SAMEORIGIN
< Content-Type: text/plain; version=0.0.4;charset=utf-8
< Transfer-Encoding: chunked
< Date: Sun, 19 Mar 2017 09:36:20 GMT
<
* Connection #0 to host 192.168.1.158 left intact
"# HELP http_req_count Number of requests\n# TYPE http_req_count gauge\nhttp_req_count{method=\"HEAD\",} 0.0\n# HELP integration_handlerCount integration_handlerCount\n# TYPE integration_handlerCount gauge\nintegration_handlerCount 0.0\n# HELP integration_channelCount integration_channelCount\n# TYPE integration_channelCount gauge\nintegration_channelCount 0.0\n# HELP integration_sourceCount integration_sourceCount\n# TYPE integration_sourceCount gauge\nintegration_sourceCount 0.0\n# HELP httpsessions_max httpsessions_max\n# TYPE httpsessions_max gauge\nhttpsessions_max -1.0\n# HELP httpsessions_active httpsessions_active\n# TYPE httpsessions_active gauge\nhttpsessions_active 567.0\n# HELP mem mem\n# TYPE mem gauge\nmem 1491693.0\n# HELP mem_free mem_free\n# TYPE mem_free gauge\nmem_free 1258272.0\n# HELP processors processors\n# TYPE processors gauge\nprocessors 8.0\n# HELP instance_uptime instance_uptime\n# TYPE instance_uptime gauge\ninstance_uptime 57678.0\n# HELP uptime uptime\n# TYPE uptime gauge\nuptime 70839.0\n# HELP systemload_average systemload_average\n# TYPE systemload_average gauge\nsystemload_average 3.07177734375\n# HELP heap_committed heap_committed\n# TYPE heap_committed gauge\nheap_committed 1382400.0\n# HELP heap_init heap_init\n# TYPE heap_init gauge\nheap_init 262144.0\n# HELP heap_used heap_used\n# TYPE heap_used gauge\nheap_used 124127.0\n# HELP heap heap\n# TYPE heap gauge\nheap 3728384.0\n# HELP nonheap_committed nonheap_committed\n# TYPE nonheap_committed gauge\nnonheap_committed 110872.0\n# HELP nonheap_init nonheap_init\n# TYPE nonheap_init gauge\nnonheap_init 2496.0\n# HELP nonheap_used nonheap_used\n# TYPE nonheap_used gauge\nnonheap_used 109293.0\n# HELP nonheap nonheap\n# TYPE nonheap gauge\nnonheap 0.0\n# HELP threads_peak threads_peak\n# TYPE threads_peak gauge\nthreads_peak 197.0\n# HELP threads_daemon threads_daemon\n# TYPE threads_daemon gauge\nthreads_daemon 188.0\n# HELP threads_totalStarted threads_totalStarted\n# TYPE threads_totalStarted gauge\nthreads_totalStarted 208.0\n# HELP threads threads\n# TYPE threads gauge\nthreads 197.0\n# HELP classes classes\n# TYPE classes gauge\nclasses 13350.0\n# HELP classes_loaded classes_loaded\n# TYPE classes_loaded gauge\nclasses_loaded 13354.0\n# HELP classes_unloaded classes_unloaded\n# TYPE classes_unloaded gauge\nclasses_unloaded 4.0\n# HELP gc_ps_scavenge_count gc_ps_scavenge_count\n# TYPE gc_ps_scavenge_count gauge\ngc_ps_scavenge_count 14.0\n# HELP gc_ps_scavenge_time gc_ps_scavenge_time\n# TYPE gc_ps_scavenge_time gauge\ngc_ps_scavenge_time 273.0\n# HELP gc_ps_marksweep_count gc_ps_marksweep_count\n# TYPE gc_ps_marksweep_count gauge\ngc_ps_marksweep_count 3.0\n# HELP gc_ps_marksweep_time gc_ps_marksweep_time\n# TYPE gc_ps_marksweep_time gauge\ngc_ps_marksweep_time 344.0\n# HELP http_req_time Time for requests\n# TYPE http_req_time histogram\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.005\",} 5.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.01\",} 13.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.025\",} 15.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.05\",} 15.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.075\",} 15.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.1\",} 15.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.25\",} 15.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.5\",} 15.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"0.75\",} 15.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"1.0\",} 15.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"2.5\",} 16.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"5.0\",} 16.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"7.5\",} 16.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"10.0\",} 16.0\nhttp_req_time_bucket{method=\"HEAD\",uri=\"/\",le=\"+Inf\",} 16.0\nhttp_req_time_count{method=\"HEAD\",uri=\"/\",} 16.0\nhttp_req_time_sum{method=\"HEAD\",uri=\"/\",} 1.3751117060000002\n"

```

<img width="1656" alt="screen shot 2017-03-19 at 11 36 02" src="https://cloud.githubusercontent.com/assets/4960306/24079760/9569e754-0c98-11e7-9c23-cdc4a858b1a5.png">
<img width="1652" alt="screen shot 2017-03-19 at 11 35 47" src="https://cloud.githubusercontent.com/assets/4960306/24079761/956e28be-0c98-11e7-9dce-9ed3cb0f59f4.png">

Prometheus:
```
Build Information
Version	1.5.2
Revision	bd1182d29f462c39544f94cc822830e1c64cf55b
Branch	master
BuildUser	root@1a01c5f68840
BuildDate	20170210-16:23:28
GoVersion	go1.7.5
```

Spring-boot client:
```
<dependency>
            <groupId>io.prometheus</groupId>
            <artifactId>simpleclient_common</artifactId>
            <version>0.0.21</version>
        </dependency>
```

**Comment by brian-brazil** [comment]
You're not exposing the result string correctly, that looks to be a literal \n rather than a newline.

Usage questions are best asked on https://groups.google.com/forum/#!aboutgroup/prometheus-users

**Comment by radut** [comment]
I have noticed that, but it is not me the one who is exporting data, the library simpleclient_spring_boot.

I noticed that I pasted only common in my first post.

**Comment by radut** [comment]
I added 
```
@EnablePrometheusEndpoint
@EnableSpringBootMetricsCollector
```
to my spring boot initializer class
and setup prometheus 


```
  - job_name: 'springboot-actuator'

    # Override the global default and scrape targets from this job every 5 seconds.
    scrape_interval: 2s
    #scrape_timeout: 3s

    metrics_path: '/actuator/prometheus'
    # scheme defaults to 'http'.

    static_configs:
         - targets: ['192.168.1.158:8083']
           labels:
             app: 'springboot'
```

**Comment by radut** [comment]
The issue seems to be because of 
```
/**
 * Specialized {@link RequestMapping} for {@link RequestMethod#GET GET} requests that
 * produce {@code application/json} or
 * {@code application/vnd.spring-boot.actuator.v1+json} responses.
 *
 * @author Andy Wilkinson
 */
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@RequestMapping(method = RequestMethod.GET, produces = {
		ActuatorMediaTypes.APPLICATION_ACTUATOR_V1_JSON_VALUE,
		MediaType.APPLICATION_JSON_VALUE })
@interface ActuatorGetMapping {

	/**
	 * Alias for {@link RequestMapping#value}.
	 * @return the value
	 */
	@AliasFor(annotation = RequestMapping.class)
	String[] value() default {};

}
```
and the response is always a json...

**Comment by radut** [comment]
If anyone else comes into this issue again
copy paste all classes from io.prometheus.client.spring.boot into a local package and then write this class next to them. Be sure to add @EnablePrometheusEndpoint from your package

```
@Component
public class PrometheusEndpointFix extends EndpointMvcAdapter {

    private final PrometheusEndpoint delegate;

    @Autowired
    public PrometheusEndpointFix(PrometheusEndpoint delegate) {
        super(delegate);
        this.delegate = delegate;
    }

    @RequestMapping(value = "", method = RequestMethod.GET, produces = MediaType.TEXT_PLAIN_VALUE)
    public ResponseEntity<String> invoke() {
        return delegate.invoke();
    }


}
```


this is because spring boot  version>=1.4.5 puts produces json by default for all endpoints

**Comment by adityagoel123** [comment]
Followed steps given below to send Metrics from Spring Boot to Prometheus:

In pom.xml added this:

```
<parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.0.0.RELEASE</version>
        <relativePath />
    </parent>

<dependency>
            <groupId>io.micrometer</groupId>
            <artifactId>micrometer-registry-prometheus</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-actuator</artifactId>
</dependency>

```

Say, we started our spring boot microservice at t=0, and then we observed the actuator-metrics. At t=0, actuator-metrics appear to be coming absolutely fine.

URL at which checked the metrics :- https://:8080/actuator/prometheus

Metric checked for :-

http_server_requests_seconds_count{application="Prod-MS",exception="None",method="POST",status="200",uri="/api/customers/createCustomer",team="ABC",} 166.0
Upon rehitting this url again within some seconds, this metric gets updated, since traffic is coming constantly !

http_server_requests_seconds_count{application="Prod-MS",exception="None",method="POST",status="200",uri="/api/customers/createCustomer",team="ABC",} 171.0
Now, at time t=3 hours or sometimes t=2.5 hours, these metrics stops getting updated and becomes constant at some random value:- At this time, application is working perfectly fine.

http_server_requests_seconds_count{application="Prod-MS",exception="None",method="POST",status="200",uri="/api/customers/createCustomer",team="ABC",} 1120.0
and it becomes constant at some random value :-

http_server_requests_seconds_count{application="Prod-MS",exception="None",method="POST",status="200",uri="/api/customers/createCustomer",team="ABC",} 1120.0
Due to this issue, we are not able to do the monitoring from Prometheus, as metrics are not coming from actuator itself.

Once the application is taken restart, the problem solves and again after 3 hours, metrics stops coming.

Following are the properties we have set :-

management.endpoints.web.exposure.include=*
management.endpoint.health.show-details=always
management.metrics.enable.http.client.requests=true
management.metrics.distribution.percentiles-histogram.http.client.requests=true
management.metrics.enable.http.server.requests=true
management.metrics.distribution.percentiles-histogram.http.server.requests=true
management.metrics.tags.tag=test
Any help shall be highly appreciated. Kindly help .
